src/lib/PerceptronPredictor.cc

```diff
@@ -0,0 +1,153 @@
+#include "simeng/PerceptronPredictor.hh"
+
+namespace simeng {
+
+PerceptronPredictor::PerceptronPredictor(ryml::ConstNodeRef config)
+    : btbBits_(config["Branch-Predictor"]["BTB-Tag-Bits"].as<uint64_t>()),
+      globalHistoryLength_(
+          config["Branch-Predictor"]["Global-History-Length"].as<uint64_t>()),
+      rasSize_(config["Branch-Predictor"]["RAS-entries"].as<uint64_t>()) {
+  // Build BTB based on config options
+  uint32_t btbSize = (1 << btbBits_);
+  btb_.resize(btbSize);
+  for (int i = 0; i < btbSize; i++) {
+    btb_[i].first.assign(globalHistoryLength_, 0);
+    btb_[i].first.push_back(1);
+    btb_[i].second = 0;
+  }
+
+  // Set up training threshold according to empirically determined formula
+  trainingThreshold_ = (int)((1.93 * (globalHistoryLength_)) + 14);
+}
+
+PerceptronPredictor::~PerceptronPredictor() {
+  ras_.clear();
+  rasHistory_.clear();
+}
+
+BranchPrediction PerceptronPredictor::predict(uint64_t address, BranchType type,
+                                              int64_t knownOffset) {
+  // Get index via an XOR hash between the global history and the lower btbBits_
+  // bits of the instruction address
+  uint64_t hashedIndex = ((address >> 2) ^ globalHistory_) & ((1 << btbBits_) - 1);
+
+  // Store the global history for correct hashing in update() --
+  // needs to be global history and not the hashed index as hashing looses information at longer
+  // global history lengths
+  btbHistory_[address] = globalHistory_;
+
+  // Retrieve the perceptron from the BTB
+  std::vector<int8_t> perceptron = btb_[hashedIndex].first;
+
+  // Determine direction prediction from perceptron by taking its dot product
+  // with the global history
+  // Starting with the bias weight
+  int64_t Pout = perceptron[globalHistoryLength_];
+  for (int i = 0; i < globalHistoryLength_; i++) {
+    bool historyTaken =
```

Comment:
Short comment explaining this line e.g. get branch direction for ith entry in the global history